rattan-core/src/cells/delay.rs

```diff
@@ -52,25 +49,75 @@ where
     P: Packet,
 {
     egress: mpsc::UnboundedReceiver<P>,
-    delay: Delay,
+    delays: Configs<Delay>,
     config_rx: mpsc::UnboundedReceiver<DelayCellConfig>,
     timer: Timer,
-    state: AtomicI32,
+    state: AtomicState,
     notify_rx: Option<tokio::sync::broadcast::Receiver<crate::control::RattanNotify>>,
     started: bool,
+    latest_egress_timestamp: Instant,
 }
 
 impl<P> DelayCellEgress<P>
 where
     P: Packet + Send + Sync,
 {
     fn set_config(&mut self, config: DelayCellConfig) {
+        let timestamp = Instant::now();
         debug!(
-            before = ?self.delay,
-            after = ?config.delay,
+            before = ?self.delays.configs.last_key_value(),
+            after = ?(timestamp, config.delay),
             "Set inner delay:"
         );
-        self.delay = config.delay;
+        self.delays.insert(timestamp, config.delay.into());
+    }
+}
+
+impl Configs<Delay> {
+    // fn delay(&self, packet_timestamp: Instant) -> Delay {
+    //     let mut packet_delay = Delay::ZERO;
+    //     // Todo: eliminate out-dated items from the Configs<Delay>
+    //     for (timestamp, delay) in self.iter() {
+    //         if *timestamp <= packet_timestamp {
+    //             packet_delay = *delay;
+    //         } else if packet_timestamp + packet_delay < *timestamp {
+    //             // If the delay applies after the packet exited the queue
+    //             // then this new delay does not apply and all after neither
+    //             break;
+    //         } else if packet_timestamp + *delay <= *timestamp {
+    //             // If the delay applies before the packet exited the queue
+    //             // but with the new delay it should have left before the new delay
+    //             // then the packet leaves immediately and no further new delay apply
+    //             packet_delay = *timestamp - packet_timestamp;
+    //             break;
+    //         } else {
+    //             packet_delay = *delay
+    //         }
+    //     }
+    //     packet_delay
+    // }
+    // ************ Which behavior here is right is under discussion. ***********
+    fn delay(&mut self, packet_timestamp: Instant) -> Delay {
+        // Eliminate out-dated items from the Configs<Delay>
+        // The correctness of this function relies on that the `pakcet_timestamp` never decreases.
+
```

Comment:
I think we shall not use this kind of get delay, which is not logically fluent. Please refer to my other comments that we shall make configuration simpler.